encodings/fastlanes/src/bitpacking/compress.rs

```diff
@@ -178,28 +179,66 @@ pub fn gather_patches(
     parray: &PrimitiveArray,
     bit_width: u8,
     num_exceptions_hint: usize,
-) -> Option<Patches> {
+) -> VortexResult<Option<Patches>> {
     let patch_validity = match parray.validity() {
         Validity::NonNullable => Validity::NonNullable,
         _ => Validity::AllValid,
     };
 
-    match_each_integer_ptype!(parray.ptype(), |$T| {
-        let mut indices: BufferMut<u64> = BufferMut::with_capacity(num_exceptions_hint);
-        let mut values: BufferMut<$T> = BufferMut::with_capacity(num_exceptions_hint);
-        for (i, v) in parray.as_slice::<$T>().iter().enumerate() {
-            if (v.leading_zeros() as usize) < parray.ptype().bit_width() - bit_width as usize && parray.is_valid(i).vortex_expect("validity") {
-                indices.push(i as u64);
-                values.push(*v);
-            }
+    let array_len = parray.len();
+    let is_valid_fn = |idx| parray.is_valid(idx).vortex_expect("validity");
```

Comment:
get the validity mask once instead of using validity

Reply:
done